data_model/src/lib.rs

```diff
@@ -1135,10 +1141,243 @@ impl From<LengthLimits> for RangeInclusive<usize> {
+                eq(
+                    &PredicateSymbol::unwrapped_and(a(), b()),
+                    &PredicateSymbol::unwrapped_and(b(), a())
+                ),
+                "Commutativity doesn't hold for `PredicateSymbol::and` implementation for {typ}"
+            );
+
+                assert!(
+                    eq(
+                        &PredicateSymbol::unwrapped_or(a(), b()),
+                        &PredicateSymbol::unwrapped_or(b(), a())
+                    ),
+                    "Commutativity doesn't hold for `PredicateSymbol::or` implementation for {typ}"
+                );
+
+                assert!(
+                    eq(
+                        &PredicateSymbol::unwrapped_or(!a(), !b()),
+                        &!PredicateSymbol::unwrapped_and(a(), b())
+                    ),
+                    "De Morgan's law doesn't hold for {typ}",
+                );
+
+                assert!(
+                    eq(
+                        &PredicateSymbol::unwrapped_and(!a(), !b()),
+                        &!PredicateSymbol::unwrapped_or(a(), b())
+                    ),
+                    "De Morgan's law doesn't hold for {typ}",
+                );
+            }
+        }
+
+        for a in &values {
+            for b in &values {
+                for c in &values {
+                    assert!(
+                    eq(
+                        &PredicateSymbol::unwrapped_and(
+                            PredicateSymbol::unwrapped_and(a(), b()),
+                            c()
+                        ),
+                        &PredicateSymbol::unwrapped_and(
+                            a(),
+                            PredicateSymbol::unwrapped_and(b(), c()),
+                        ),
+                    ),
+                    "Associativity doesn't hold for `PredicateSymbol::or` implementation for {typ}",
+                );
+
+                    assert!(
+                    eq(
+                        &PredicateSymbol::unwrapped_or(
+                            PredicateSymbol::unwrapped_or(a(), b()),
+                            c()
+                        ),
+                        &PredicateSymbol::unwrapped_or(
+                            a(),
+                            PredicateSymbol::unwrapped_or(b(), c()),
+                        ),
+                    ),
+                    "Associativity doesn't hold for `PredicateSymbol::and` implementation for {typ}",
+                );
+                }
+            }
+        }
+    }
+}
+
+impl PredicateSymbol for bool {
+    fn and(self, other: Self) -> ControlFlow<Self, Self> {
+        if self && other {
+            ControlFlow::Continue(true)
+        } else {
+            ControlFlow::Break(false)
+        }
+    }
+
+    fn or(self, other: Self) -> ControlFlow<Self, Self> {
+        if self || other {
+            ControlFlow::Break(true)
+        } else {
+            ControlFlow::Continue(false)
+        }
+    }
+}
```

Comment:
While normally the teranry ⇒ binary monadic correspondence is written with an `Option` or a monoid like `Result`, I like this representation. It's more vivid and more explicit about the interaction of different behaviours. I'm glad you decided to rewrite it this way.